File: python/aigov_py/audit_export_signing.py
"""
Ed25519 signing and verification for ``aigov.audit_export.v1`` documents.

Signing is a post-export step: it does not modify append-only ledger semantics.
The canonical signing payload binds schema version, run_id, content hashes, and verdict.
"""

from __future__ import annotations

import base64
import hashlib
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterable

from nacl.signing import SigningKey, VerifyKey

from aigov_py.canonical_json import canonical_bytes
from aigov_py.policy_bundle_signing import TrustEd25519, _iter_verify_keys, load_trust_from_env_json
from aigov_py.portable_evidence_digest import portable_evidence_digest_v1

SUPPORTED_EXPORT_SCHEMA = "aigov.audit_export.v1"
SIGNATURE_KIND_ED25519 = "ed25519"


@dataclass(frozen=True)
class AuditExportSigningPayload:
    """Deterministic preimage fields for an export signature (excludes ``signatures``)."""

    schema_version: str
    run_id: str
    policy_version: str
    environment: str
    bundle_sha256: str
    events_content_sha256: str
    chain_head_record_sha256: str
    decision_verdict: str

    def as_dict(self) -> dict[str, str]:
        return {
            "schema_version": self.schema_version,
            "run_id": self.run_id,
            "policy_version": self.policy_version,
            "environment": self.environment,
            "bundle_sha256": self.bundle_sha256,
            "events_content_sha256": self.events_content_sha256,
            "chain_head_record_sha256": self.chain_head_record_sha256,
            "decision_verdict": self.decision_verdict,
        }


def export_run_id(export_doc: dict[str, Any]) -> str:
    run = export_doc.get("run")
    if not isinstance(run, dict):
        return ""
    return str(run.get("run_id") or "").strip()


def canonical_audit_export_signing_payload(export_doc: dict[str, Any]) -> AuditExportSigningPayload:
    """Build the deterministic preimage used for Ed25519 signing and verification."""

    doc = dict(export_doc)
    doc.pop("signatures", None)
    eh = doc.get("evidence_hashes")
    if not isinstance(eh, dict):
        eh = {}
    decision = doc.get("decision")
    if not isinstance(decision, dict):
        decision = {}
    chain_head = eh.get("chain_head_record_sha256")
    return AuditExportSigningPayload(
        schema_version=str(doc.get("schema_version") or "").strip(),
        run_id=export_run_id(doc),
        policy_version=str(doc.get("policy_version") or "").strip(),
        environment=str(doc.get("environment") or "").strip(),
        bundle_sha256=str(eh.get("bundle_sha256") or "").strip().lower(),
        events_content_sha256=str(eh.get("events_content_sha256") or "").strip().lower(),
        chain_head_record_sha256=str(chain_head or "").strip().lower()
        if chain_head is not None
        else "",
        decision_verdict=str(decision.get("verdict") or "").strip(),
    )


def audit_export_payload_digest_sha256_hex(export_doc: dict[str, Any]) -> str:
    """SHA-256 hex digest over canonical JSON of the signing payload."""

    payload = canonical_audit_export_signing_payload(export_doc)
    return hashlib.sha256(canonical_bytes(payload.as_dict())).hexdigest()


def _require_sha256_hex(label: str, value: str) -> str:
    v = (value or "").strip().lower()
    if len(v) != 64 or not all(c in "0123456789abcdef" for c in v):
        raise ValueError(f"{label} must be 64 lowercase hex characters")
    return v


def validate_export_schema_version(export_doc: dict[str, Any]) -> None:
    schema = str(export_doc.get("schema_version") or "").strip()
    if schema != SUPPORTED_EXPORT_SCHEMA:
        raise ValueError(
            f"unsupported export schema {schema!r} (expected {SUPPORTED_EXPORT_SCHEMA})"
        )


def validate_export_run_id_consistency(export_doc: dict[str, Any]) -> str:
    run_id = export_run_id(export_doc)
    if not run_id:
        raise ValueError("run.run_id is missing or empty")
    events = export_doc.get("evidence_events")
    if not isinstance(events, list):
        raise ValueError("evidence_events must be an array")
    for i, ev in enumerate(events):
        if not isinstance(ev, dict):
            raise ValueError(f"evidence_events[{i}] must be an object")
        ev_run = str(ev.get("run_id") or "").strip()
        if ev_run != run_id:
            raise ValueError(
                f"evidence_events[{i}].run_id {ev_run!r} inconsistent with run.run_id {run_id!r}"
            )
    return run_id


def validate_events_content_sha256(export_doc: dict[str, Any], *, run_id: str) -> None:
    eh = export_doc.get("evidence_hashes")
    if not isinstance(eh, dict):
        raise ValueError("evidence_hashes must be an object")
    declared = _require_sha256_hex(
        "evidence_hashes.events_content_sha256",
        str(eh.get("events_content_sha256") or ""),
    )
    events = export_doc.get("evidence_events")
    if not isinstance(events, list):
        raise ValueError("evidence_events must be an array")
    recomputed = portable_evidence_digest_v1(run_id, [dict(e) for e in events if isinstance(e, dict)])
    if recomputed != declared:
        raise ValueError(
            "events_content_sha256 mismatch: export evidence does not match recomputed digest"
        )


def validate_bundle_sha256_field(export_doc: dict[str, Any]) -> str:
    eh = export_doc.get("evidence_hashes")
    if not isinstance(eh, dict):
        raise ValueError("evidence_hashes must be an object")
    return _require_sha256_hex(
        "evidence_hashes.bundle_sha256",
        str(eh.get("bundle_sha256") or ""),
    )


def validate_signed_audit_export_integrity(export_doc: dict[str, Any]) -> AuditExportSigningPayload:
    """
    Structural and digest checks before signature verification.

    Does not verify Ed25519 signatures; use :func:`verify_signed_audit_export`.
    """

    validate_export_schema_version(export_doc)
    run_id = validate_export_run_id_consistency(export_doc)
    validate_events_content_sha256(export_doc, run_id=run_id)
    validate_bundle_sha256_field(export_doc)
    payload = canonical_audit_export_signing_payload(export_doc)
    if payload.run_id != run_id:
        raise ValueError("canonical signing payload run_id mismatch")
    return payload


def sign_audit_export_ed25519(
    export_path: str | Path,
    *,
    out_path: str | Path,
    issuer_id: str,
    signer: str,
    private_key_base64: str,
    created_at_utc: str,
    expires_at_utc: str | None = None,
) -> dict[str, Any]:
    """Sign an audit export JSON file and write the document with a ``signatures`` array."""

    p = Path(export_path)
    doc = json.loads(p.read_text(encoding="utf-8"))
    if not isinstance(doc, dict):
        raise ValueError("audit export must be a JSON object")
    validate_signed_audit_export_integrity(doc)

    digest = audit_export_payload_digest_sha256_hex(doc)
    sk = SigningKey(base64.b64decode(private_key_base64.strip()))
    sig_bytes = sk.sign(digest.encode("utf-8")).signature
    sig_b64 = base64.b64encode(sig_bytes).decode("ascii")
    payload = canonical_audit_export_signing_payload(doc)

    sig_record: dict[str, Any] = {
        "kind": SIGNATURE_KIND_ED25519,
        "issuer_id": issuer_id.strip(),
        "signer": signer.strip(),
        "created_at_utc": created_at_utc,
        "expires_at_utc": expires_at_utc,
        "payload_digest_sha256": digest,
        "canonical_payload": payload.as_dict(),
        "signature": {"encoding": "base64", "value": sig_b64},
    }

    sigs = doc.get("signatures")
    if sigs is None:
        doc["signatures"] = [sig_record]
    elif isinstance(sigs, list):
        doc["signatures"] = list(sigs) + [sig_record]
    else:
        raise ValueError("export.signatures must be an array when present")

    out = Path(out_path)
    out.write_text(json.dumps(doc, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    return doc


def verify_audit_export_ed25519_signature(
    export_doc: dict[str, Any],
    *,
    trust: Iterable[TrustEd25519],
) -> None:
    """Verify at least one valid Ed25519 signature on the export canonical digest."""

    validate_signed_audit_export_integrity(export_doc)
    digest = audit_export_payload_digest_sha256_hex(export_doc)

    sigs = export_doc.get("signatures")
    if not isinstance(sigs, list) or not sigs:
        raise ValueError("export is unsigned (signatures missing or empty)")

    last_err: str | None = None
    for s in sigs:
        if not isinstance(s, dict):
            last_err = "signature record must be an object"
            continue
        if str(s.get("kind") or "") != SIGNATURE_KIND_ED25519:
            last_err = "unsupported signature kind"
            continue
        if str(s.get("payload_digest_sha256") or "").strip().lower() != digest:
            last_err = "payload digest mismatch"
            continue
        issuer_id = str(s.get("issuer_id") or "").strip()
        if not issuer_id:
            last_err = "signature issuer_id missing"
            continue
        sig = s.get("signature")
        if not isinstance(sig, dict) or str(sig.get("encoding") or "") != "base64":
            last_err = "unsupported signature encoding"
            continue
        sig_b64 = str(sig.get("value") or "").strip()
        if not sig_b64:
            last_err = "signature value empty"
            continue
        try:
            sig_bytes = base64.b64decode(sig_b64)
        except Exception as e:  # noqa: BLE001
            last_err = f"invalid signature base64: {e}"
            continue

        ok = False
        for vk in _iter_verify_keys(trust, issuer_id):
            try:
                vk.verify(digest.encode("utf-8"), sig_bytes)
                ok = True
                break
            except Exception as e:  # noqa: BLE001
                last_err = str(e) or "verify failed"
        if ok:
            return
    raise ValueError(last_err or "no valid export signature found")


def verify_signed_audit_export(
    export_doc: dict[str, Any],
    *,
    trust: Iterable[TrustEd25519],
) -> AuditExportSigningPayload:
    """Full verification: integrity checks + Ed25519 signature."""

    verify_audit_export_ed25519_signature(export_doc, trust=trust)
    return canonical_audit_export_signing_payload(export_doc)


__all__ = [
    "SUPPORTED_EXPORT_SCHEMA",
    "SIGNATURE_KIND_ED25519",
    "AuditExportSigningPayload",
    "TrustEd25519",
    "load_trust_from_env_json",
    "export_run_id",
    "canonical_audit_export_signing_payload",
    "audit_export_payload_digest_sha256_hex",
    "validate_export_schema_version",
    "validate_export_run_id_consistency",
    "validate_events_content_sha256",
    "validate_bundle_sha256_field",
    "validate_signed_audit_export_integrity",
    "sign_audit_export_ed25519",
    "verify_audit_export_ed25519_signature",
    "verify_signed_audit_export",
]

File: rust/src/audit_export_signing.rs
//! Ed25519 signing and verification for ``aigov.audit_export.v1`` documents.

use crate::bundle::portable_evidence_digest_v1;
use crate::canonical_json::{canonical_json_bytes, sha256_hex_bytes, sort_json_value};
use crate::policy_signing::PolicyTrustStore;
use crate::replay_validation::{export_run_id, EXPORT_SCHEMA_V1};
use crate::schema::EvidenceEvent;
use base64::Engine;
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

pub const SIGNATURE_KIND_ED25519: &str = "ed25519";

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct AuditExportSigningPayload {
    pub schema_version: String,
    pub run_id: String,
    pub policy_version: String,
    pub environment: String,
    pub bundle_sha256: String,
    pub events_content_sha256: String,
    pub chain_head_record_sha256: String,
    pub decision_verdict: String,
}

pub fn canonical_audit_export_signing_payload(export: &Value) -> AuditExportSigningPayload {
    let mut doc = export.clone();
    if let Some(obj) = doc.as_object_mut() {
        obj.remove("signatures");
    }
    let eh = doc
        .get("evidence_hashes")
        .and_then(|v| v.as_object())
        .cloned()
        .unwrap_or_default();
    let decision = doc
        .get("decision")
        .and_then(|v| v.as_object())
        .cloned()
        .unwrap_or_default();
    let chain_head = eh
        .get("chain_head_record_sha256")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    AuditExportSigningPayload {
        schema_version: doc
            .get("schema_version")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string(),
        run_id: export_run_id(&doc).unwrap_or_default(),
        policy_version: doc
            .get("policy_version")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string(),
        environment: doc
            .get("environment")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string(),
        bundle_sha256: eh
            .get("bundle_sha256")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_ascii_lowercase(),
        events_content_sha256: eh
            .get("events_content_sha256")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_ascii_lowercase(),
        chain_head_record_sha256: chain_head.trim().to_ascii_lowercase(),
        decision_verdict: decision
            .get("verdict")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string(),
    }
}

pub fn audit_export_payload_digest_sha256_hex(export: &Value) -> String {
    let payload = canonical_audit_export_signing_payload(export);
    sha256_hex_bytes(&canonical_json_bytes(&payload))
}

fn require_sha256_hex(label: &str, value: &str) -> Result<String, String> {
    let v = value.trim().to_ascii_lowercase();
    if v.len() != 64 || !v.chars().all(|c| c.is_ascii_hexdigit()) {
        return Err(format!("{label} must be 64 lowercase hex characters"));
    }
    Ok(v)
}

pub fn validate_export_schema_version(export: &Value) -> Result<(), String> {
    let schema = export
        .get("schema_version")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim();
    if schema != EXPORT_SCHEMA_V1 {
        return Err(format!(
            "unsupported export schema {schema:?} (expected {EXPORT_SCHEMA_V1})"
        ));
    }
    Ok(())
}

pub fn validate_export_run_id_consistency(export: &Value) -> Result<String, String> {
    let run_id = export_run_id(export).ok_or_else(|| "run.run_id is missing or empty".to_string())?;
    let events = export
        .get("evidence_events")
        .and_then(|v| v.as_array())
        .ok_or_else(|| "evidence_events must be an array".to_string())?;
    for (i, ev) in events.iter().enumerate() {
        let ev_run = ev
            .get("run_id")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim();
        if ev_run != run_id {
            return Err(format!(
                "evidence_events[{i}].run_id {ev_run:?} inconsistent with run.run_id {run_id:?}"
            ));
        }
    }
    Ok(run_id)
}

pub fn validate_events_content_sha256(export: &Value, run_id: &str) -> Result<(), String> {
    let eh = export
        .get("evidence_hashes")
        .and_then(|v| v.as_object())
        .ok_or_else(|| "evidence_hashes must be an object".to_string())?;
    let declared = require_sha256_hex(
        "evidence_hashes.events_content_sha256",
        eh.get("events_content_sha256")
            .and_then(|v| v.as_str())
            .unwrap_or(""),
    )?;
    let events = export
        .get("evidence_events")
        .and_then(|v| v.as_array())
        .ok_or_else(|| "evidence_events must be an array".to_string())?;
    let mut parsed: Vec<EvidenceEvent> = Vec::with_capacity(events.len());
    for (i, item) in events.iter().enumerate() {
        let ev: EvidenceEvent = serde_json::from_value(item.clone())
            .map_err(|e| format!("evidence_events[{i}] invalid: {e}"))?;
        parsed.push(ev);
    }
    let recomputed = portable_evidence_digest_v1(run_id, &parsed);
    if recomputed != declared {
        return Err(
            "events_content_sha256 mismatch: export evidence does not match recomputed digest"
                .to_string(),
        );
    }
    Ok(())
}

pub fn validate_bundle_sha256_field(export: &Value) -> Result<String, String> {
    let eh = export
        .get("evidence_hashes")
        .and_then(|v| v.as_object())
        .ok_or_else(|| "evidence_hashes must be an object".to_string())?;
    require_sha256_hex(
        "evidence_hashes.bundle_sha256",
        eh.get("bundle_sha256")
            .and_then(|v| v.as_str())
            .unwrap_or(""),
    )
}

pub fn validate_signed_audit_export_integrity(export: &Value) -> Result<AuditExportSigningPayload, String> {
    validate_export_schema_version(export)?;
    let run_id = validate_export_run_id_consistency(export)?;
    validate_events_content_sha256(export, &run_id)?;
    validate_bundle_sha256_field(export)?;
    let payload = canonical_audit_export_signing_payload(export);
    if payload.run_id != run_id {
        return Err("canonical signing payload run_id mismatch".to_string());
    }
    Ok(payload)
}

pub fn verify_audit_export_ed25519_signature(
    export: &Value,
    trust: &PolicyTrustStore,
    expected_issuer_id: Option<&str>,
) -> Result<String, String> {
    validate_signed_audit_export_integrity(export)?;
    let digest = audit_export_payload_digest_sha256_hex(export);

    let sigs = export
        .get("signatures")
        .and_then(|v| v.as_array())
        .filter(|a| !a.is_empty())
        .ok_or_else(|| "export is unsigned (signatures missing or empty)".to_string())?;

    let mut last_err = "no valid export signature found".to_string();
    for s in sigs {
        let Some(sig_obj) = s.as_object() else {
            last_err = "signature record must be an object".to_string();
            continue;
        };
        if sig_obj
            .get("kind")
            .and_then(|v| v.as_str())
            != Some(SIGNATURE_KIND_ED25519)
        {
            last_err = "unsupported signature kind".to_string();
            continue;
        }
        if sig_obj
            .get("payload_digest_sha256")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_ascii_lowercase()
            != digest
        {
            last_err = "payload digest mismatch".to_string();
            continue;
        }
        let issuer_id = sig_obj
            .get("issuer_id")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim();
        if issuer_id.is_empty() {
            last_err = "signature issuer_id missing".to_string();
            continue;
        }
        if let Some(expected) = expected_issuer_id {
            if expected.trim() != issuer_id {
                last_err = format!("signature issuer_id {issuer_id} does not match expected {expected}");
                continue;
            }
        }
        let sig = sig_obj.get("signature").and_then(|v| v.as_object());
        let Some(sig) = sig else {
            last_err = "signature object missing".to_string();
            continue;
        };
        if sig.get("encoding").and_then(|v| v.as_str()) != Some("base64") {
            last_err = "unsupported signature encoding".to_string();
            continue;
        }
        let sig_b64 = sig.get("value").and_then(|v| v.as_str()).unwrap_or("").trim();
        if sig_b64.is_empty() {
            last_err = "signature value empty".to_string();
            continue;
        }
        let sig_bytes = base64::engine::general_purpose::STANDARD
            .decode(sig_b64)
            .map_err(|e| format!("invalid signature base64: {e}"))?;
        let signature = Signature::from_slice(sig_bytes.as_slice())
            .map_err(|e| format!("invalid ed25519 signature bytes: {e}"))?;

        let keys = match trust.ed25519_pubkeys.get(issuer_id) {
            Some(k) => k,
            None => {
                last_err = format!("no trusted keys for issuer_id={issuer_id}");
                continue;
            }
        };
        for vk in keys {
            if vk.verify(digest.as_bytes(), &signature).is_ok() {
                return Ok(issuer_id.to_string());
            }
            last_err = "ed25519 verify failed".to_string();
        }
    }
    Err(last_err)
}

pub fn sign_audit_export_ed25519(
    export: &mut Value,
    issuer_id: &str,
    signer: &str,
    signing_key: &SigningKey,
    created_at_utc: &str,
    expires_at_utc: Option<&str>,
) -> Result<(), String> {
    validate_signed_audit_export_integrity(export)?;
    let digest = audit_export_payload_digest_sha256_hex(export);
    let payload = canonical_audit_export_signing_payload(export);
    let sig_bytes = signing_key.sign(digest.as_bytes());
    let sig_b64 = base64::engine::general_purpose::STANDARD.encode(sig_bytes.to_bytes());

    let sig_record = json!({
        "kind": SIGNATURE_KIND_ED25519,
        "issuer_id": issuer_id.trim(),
        "signer": signer.trim(),
        "created_at_utc": created_at_utc,
        "expires_at_utc": expires_at_utc,
        "payload_digest_sha256": digest,
        "canonical_payload": sort_json_value(serde_json::to_value(&payload).expect("payload json")),
        "signature": {"encoding": "base64", "value": sig_b64},
    });

    match export.get_mut("signatures") {
        None => {
            export
                .as_object_mut()
                .expect("export object")
                .insert("signatures".to_string(), json!([sig_record]));
        }
        Some(v) => {
            let arr = v
                .as_array_mut()
                .ok_or_else(|| "export.signatures must be an array when present".to_string())?;
            arr.push(sig_record);
        }
    }
    Ok(())
}

pub fn signing_key_from_seed(seed: &[u8; 32]) -> SigningKey {
    SigningKey::from_bytes(seed)
}

pub fn verifying_key_from_seed(seed: &[u8; 32]) -> VerifyingKey {
    SigningKey::from_bytes(seed).verifying_key()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::policy_signing::PolicyTrustStore;
    use std::collections::BTreeMap;

    fn minimal_export(run_id: &str) -> Value {
        let events = json!([{
            "event_id": format!("{run_id}-e1"),
            "event_type": "ai_discovery_reported",
            "ts_utc": "2026-01-01T00:00:01Z",
            "actor": "test",
            "system": "test",
            "run_id": run_id,
            "payload": {"openai": false, "transformers": false, "model_artifacts": false}
        }]);
        let events_vec: Vec<EvidenceEvent> =
            serde_json::from_value(events.clone()).expect("events");
        let events_sha = portable_evidence_digest_v1(run_id, &events_vec);
        json!({
            "ok": true,
            "schema_version": EXPORT_SCHEMA_V1,
            "policy_version": "test-policy",
            "environment": "dev",
            "run": {"run_id": run_id},
            "evidence_hashes": {
                "bundle_sha256": "a".repeat(64),
                "events_content_sha256": events_sha,
                "chain_head_record_sha256": "b".repeat(64),
                "log_chain": []
            },
            "decision": {"verdict": "BLOCKED"},
            "evidence_events": events
        })
    }

    fn trust_for_seed(seed: [u8; 32]) -> PolicyTrustStore {
        let vk = verifying_key_from_seed(&seed);
        let pk_b64 = base64::engine::general_purpose::STANDARD.encode(vk.as_bytes());
        let mut map = BTreeMap::new();
        map.insert("govai-export-signer".to_string(), vec![vk]);
        let _ = pk_b64;
        PolicyTrustStore {
            ed25519_pubkeys: map,
        }
    }

    #[test]
    fn sign_and_verify_roundtrip() {
        let seed = [0x22; 32];
        let sk = signing_key_from_seed(&seed);
        let trust = trust_for_seed(seed);
        let mut export = minimal_export("run-sign-test");
        sign_audit_export_ed25519(
            &mut export,
            "govai-export-signer",
            "test",
            &sk,
            "2026-01-01T00:00:00Z",
            None,
        )
        .expect("sign");
        let issuer = verify_audit_export_ed25519_signature(&export, &trust, None).expect("verify");
        assert_eq!(issuer, "govai-export-signer");
    }
}
